File: week_1/analysis_of_algorithms/three_sum_binary_search.py
import unittest
import timeit

class ThreeSumBinarySearch:

    def __init__(self, array):
      self.array = array
      self.array.sort()
      self.n = len(self.array)

    def count(self):
      count = 0
      for i in range(self.n):
        for j in range(i+1, self.n):
          val = (self.array[i] + self.array[j]) * -1
          if self.find(j+1, self.n-1, val):
            count += 1 
      return count

    def find(self, lo, hi, key):

        while (lo <= hi):
          mid = lo + (hi - lo)/2
          if key < self.array[mid]:
            hi = mid -1
          elif key > self.array[mid]:
            lo = mid +1
          else:
            return True 

        return False


class ThreeSumBinarySearchTest(unittest.TestCase):
    
  def test_count(self):
    arr1 = [-3, -2, -1, 0, 1, 2, 3]
    obj = ThreeSumBinarySearch(arr1)
    self.assertEqual(obj.count(), 5)

    arr2 = [3, 2, 1, 0, 1, 2, 3]
    obj = ThreeSumBinarySearch(arr2)
    self.assertEqual(obj.count(), 0)

if __name__ == '__main__':
    setup = '''
from __main__ import ThreeSumBinarySearch
import random

array = random.sample(range(-1000, 1000), 100)
'''
    print(max(timeit.repeat("ThreeSumBinarySearch(array).count()", setup= setup, number= 10)))
    unittest.main()

File: week_1/analysis_of_algorithms/three_sum_binary_search.rb
require 'test/unit'
require 'benchmark'

class ThreeSumBinarySearch

  def initialize array=[]
    @array = array.sort
    @n = (@array.size) -1
  end

  def count
    count = 0
    (0..@n).to_a.each do |i|
      j = i+1
      (j..@n).to_a.each do |j|
        val = (@array[i] + @array[j]) * -1
        k = j+1
        count += 1 if search(k, @n, val)
        end
      end
    return count
  end

  def search lo, hi, key

    while (lo <= hi) do
      mid = lo + (hi - lo)/2

      if key < @array[mid]
        hi = mid -1
      elsif key > @array[mid]
        lo = mid +1
      else
        return true
      end
    end

    return false
  end

end

# Check performance
BM_ARRAY = (-10000..10000).map { rand }

Benchmark.bm do |x|
  x.report { 50.times do;  ThreeSumBinarySearch.new(BM_ARRAY.sample(100)).count(); end }
end

# Test object
class ThreeSumBinarySearchTest < Test::Unit::TestCase

  def test_count
    arr1 = [-3, -2, -1, 0, 1, 2, 3]
    obj = ThreeSumBinarySearch.new(arr1)
    assert_equal(5, obj.count())

    arr2 = [3, 2, 1, 0, 1, 2, 3]
    obj = ThreeSumBinarySearch.new(arr2)
    assert_equal(0, obj.count())
  end

end
